core/src/main/java/org/projectnessie/cel/interpreter/AttributeFactory.java

```diff
@@ -353,34 +354,35 @@ public Object resolve(org.projectnessie.cel.interpreter.Activation vars) {
      * type, then the result is `nil`, `false`, `nil` per the interface requirement.
      */
     @Override
-    public Object tryResolve(org.projectnessie.cel.interpreter.Activation vars) {
+    public Pair<Object, Boolean> tryResolve(org.projectnessie.cel.interpreter.Activation vars) {
       for (String nm : namespaceNames) {
         // If the variable is found, process it. Otherwise, wait until the checks to
         // determine whether the type is unknown before returning.
-        Object op = vars.resolveName(nm);
-        if (op != null) {
+        Pair<Object, Boolean> obj = vars.resolveName(nm);
+        if (obj.b) {
+          Object op = obj.a;
           for (Qualifier qual : qualifiers) {
             Object op2 = qual.qualify(vars, op);
             if (op2 instanceof Err) {
-              return op2;
+              return new Pair<>(op2, true);
             }
             if (op2 == null) {
               break;
             }
             op = op2;
           }
-          return op;
+          return new Pair<>(op, true);
         }
         // Attempt to resolve the qualified type name if the name is not a variable identifier.
         Val typ = provider.findIdent(nm);
         if (typ != null) {
           if (qualifiers.isEmpty()) {
-            return typ;
+            return new Pair<>(typ, true);
           }
           throw noSuchAttributeException(this);
         }
       }
-      return null;
+      return new Pair<>(null, false);
```

Comment:
Instead of returning a value, why not simply `throw noSuchAttributeException(this)` here vs waiting until `resolve`?  This way, `tryResolve` either throws or returns a (potentially null) value. In the latter case you no longer need to distinguish between absence and null.